src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,6 +453,210 @@ private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, I
+                var localsBuilder = ArrayBuilder<LocalSymbol>.GetInstance();
+
+                foreach (var arg in args)
+                {
+                    Debug.Assert(arg.Type is not null);
+
+                    if (arg.Type.IsReadOnlySpanChar())
+                    {
+                        preparedArgsBuilder.Add(arg);
+                    }
+                    else if (arg.Type.SpecialType == SpecialType.System_Char)
+                    {
+                        Debug.Assert(readOnlySpanCtorRefParamChar is not null);
+
+                        var temp = factory.StoreToTemp(arg, out var tempAssignment);
+                        localsBuilder.Add(temp.LocalSymbol);
+
+                        var wrappedChar = new BoundObjectCreationExpression(
+                            arg.Syntax,
+                            readOnlySpanCtorRefParamChar,
+                            [temp],
+                            argumentNamesOpt: default,
+                            argumentRefKindsOpt: [RefKindExtensions.StrictIn],
+                            expanded: false,
+                            argsToParamsOpt: default,
+                            defaultArguments: default,
+                            constantValueOpt: null,
+                            initializerExpressionOpt: null,
+                            type: readOnlySpanCtorRefParamChar.ContainingType);
+
+                        preparedArgsBuilder.Add(new BoundSequence(
+                            arg.Syntax,
+                            [],
+                            [tempAssignment],
+                            wrappedChar,
+                            wrappedChar.Type)
+                        { WasCompilerGenerated = true });
+                    }
+                    else
+                    {
+                        Debug.Assert(arg.HasAnyErrors || arg.Type.SpecialType == SpecialType.System_String);
+                        Debug.Assert(stringImplicitConversionToReadOnlySpan is not null);
+                        preparedArgsBuilder.Add(BoundCall.Synthesized(arg.Syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringImplicitConversionToReadOnlySpan, arg));
+                    }
+                }
+
+                var concatCall = BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, spanConcat, preparedArgsBuilder.ToImmutableAndFree());
+
+                var oldSyntax = factory.Syntax;
+                factory.Syntax = syntax;
+
+                var sequence = factory.Sequence(
+                    localsBuilder.ToImmutableAndFree(),
+                    [],
+                    concatCall);
+
+                factory.Syntax = oldSyntax;
+                return sequence;
+            }
+        }
+
+        private static WellKnownMember? GetSpanConcatMemberByArgumentsCount(int argumentCount) => argumentCount switch
```

Comment:
> => argumentCount switch [](http://example.com/codeflow?start=94&length=24)

If an expression body doesn't fit on the same line, I think the benefit of using this syntactic form disappears, and in fact the form starts negatively affect readability of the code. It takes more time for a reader to understand what is going on. Consider switching to block body.